test: Use TestNode datadir_path or chain_path where possible

maflcko opened the pull request:
It seems inconsistent, fragile and verbose to:

* Call `get_datadir_path` to recreate the path that already exists as field in TestNode
* Call `os.path.join` with the hardcoded chain name or `self.chain` to recreate the TestNode `chain_path` property
* Sometimes even use the hardcoded node dir name (`"node0"`)

Fix all issues by using the TestNode properties.

Commit:
test: Use wallet_dir lambda in wallet_multiwallet test where possible

Seems odd to hardcode all parent directory names in the path for no good
reason.

Also, add wallet_path property to TestNode.

Also, rework wallet_backup.py test for scripted-diff in the next commit.

Commit:
test: Use TestNode *_path properties where possible

Seems odd to place the burden on test writers to hardcode the chain or
datadir path for the nodes under test.

theStack reviewed (commented):
Concept ACK

Here's a few more instances which I think could be simplified by using `chain_path` (found via `$ git grep join.*regtest ./test`):

https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_backwards_compatibility.py#L75-L78
https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_backwards_compatibility.py#L160-L163
https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_inactive_hdchains.py#L133-L134
https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_migration.py#L37
https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_migration.py#L460
https://github.com/bitcoin/bitcoin/blob/1ecdf6ea8fb6dc1d64f41323a1e337beb174c25b/test/functional/wallet_upgradewallet.py#L141-L145

willcl-ark commented:
Concept ACK.

In addition to those highlighted by TheStack, there are also quite a few other places we could update to use this, e.g. some of these matches:

```sh
$ git grep -n "os.path" test/functional | rg -e '.*\.chain*'
test/functional/feature_addrman.py:56:        peers_dat = os.path.join(self.nodes[0].datadir, self.chain, "peers.dat")
test/functional/feature_asmap.py:116:        self.datadir = os.path.join(self.node.datadir, self.chain)
test/functional/feature_blocksdir.py:33:        assert os.path.isfile(os.path.join(blocksdir_path, self.chain, "blocks", "blk00000.dat"))
test/functional/feature_blocksdir.py:34:        assert os.path.isdir(os.path.join(self.nodes[0].datadir, self.chain, "blocks", "index"))
test/functional/feature_config_args.py:418:        assert os.path.exists(os.path.join(new_data_dir, self.chain, 'blocks'))
test/functional/feature_config_args.py:424:        assert os.path.exists(os.path.join(new_data_dir_2, self.chain, 'blocks'))
test/functional/feature_fee_estimation.py:323:        fee_dat = os.path.join(self.nodes[0].datadir, self.chain, "fee_estimates.dat")
test/functional/feature_filelock.py:27:        datadir = os.path.join(self.nodes[0].datadir, self.chain)
test/functional/feature_loadblock.py:40:        blocks_dir = os.path.join(data_dir, self.chain, "blocks")
test/functional/feature_logging.py:19:        return os.path.join(self.nodes[0].datadir, self.chain, name)
test/functional/feature_posix_fs_permissions.py:34:        datadir = os.path.join(self.nodes[0].datadir, self.chain)
test/functional/feature_pruning.py:94:        self.prunedir = os.path.join(self.nodes[2].datadir, self.chain, 'blocks', '')
test/functional/feature_pruning.py:293:            return os.path.isfile(os.path.join(self.nodes[node_number].datadir, self.chain, "blocks", f"blk{index:05}.dat"))
test/functional/feature_reindex.py:42:        blk0 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00000.dat')
test/functional/feature_remove_pruned_files_on_startup.py:23:        blk0 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00000.dat')
test/functional/feature_remove_pruned_files_on_startup.py:24:        rev0 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'rev00000.dat')
test/functional/feature_remove_pruned_files_on_startup.py:25:        blk1 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'blk00001.dat')
test/functional/feature_remove_pruned_files_on_startup.py:26:        rev1 = os.path.join(self.nodes[0].datadir, self.nodes[0].chain, 'blocks', 'rev00001.dat')
test/functional/feature_txindex_compatibility.py:53:        legacy_chain_dir = os.path.join(self.nodes[0].datadir, self.chain)
test/functional/feature_txindex_compatibility.py:56:        migrate_chain_dir = os.path.join(self.nodes[2].datadir, self.chain)
test/functional/feature_txindex_compatibility.py:67:        drop_index_chain_dir = os.path.join(self.nodes[1].datadir, self.chain)
test/functional/interface_rpc.py:49:        assert_equal(info['logpath'], os.path.join(self.nodes[0].datadir, self.chain, 'debug.log'))
test/functional/mempool_compatibility.py:58:        old_node_mempool = os.path.join(old_node.datadir, self.chain, 'mempool.dat')
test/functional/mempool_compatibility.py:59:        new_node_mempool = os.path.join(new_node.datadir, self.chain, 'mempool.dat')
test/functional/mempool_persist.py:146:        mempooldat0 = os.path.join(self.nodes[0].datadir, self.chain, 'mempool.dat')
test/functional/mempool_persist.py:147:        mempooldat1 = os.path.join(self.nodes[1].datadir, self.chain, 'mempool.dat')
test/functional/test_framework/test_framework.py:826:                return os.path.join(cache_node_dir, self.chain, *paths)
test/functional/tool_wallet.py:404:        self.wallet_path = os.path.join(self.nodes[0].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename)
test/functional/wallet_backup.py:112:        os.remove(os.path.join(self.nodes[0].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename))
test/functional/wallet_backup.py:113:        os.remove(os.path.join(self.nodes[1].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename))
test/functional/wallet_backup.py:114:        os.remove(os.path.join(self.nodes[2].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename))
test/functional/wallet_backup.py:121:        not_created_wallet_file = os.path.join(node.datadir, self.chain, 'wallets', wallet_name)
test/functional/wallet_backup.py:131:        not_created_wallet_file = os.path.join(node.datadir, self.chain, 'wallets', wallet_name)
test/functional/wallet_backup.py:138:        wallet_file = os.path.join(node.datadir, self.chain, 'wallets', wallet_name)
test/functional/wallet_backup.py:209:        assert os.path.exists(os.path.join(self.nodes[3].datadir, self.chain, 'wallets', "res0"))
test/functional/wallet_backup.py:210:        assert os.path.exists(os.path.join(self.nodes[3].datadir, self.chain, 'wallets', "res1"))
test/functional/wallet_backup.py:211:        assert os.path.exists(os.path.join(self.nodes[3].datadir, self.chain, 'wallets', "res2"))
test/functional/wallet_backup.py:229:            shutil.rmtree(os.path.join(self.nodes[2].datadir, self.chain, 'blocks'))
test/functional/wallet_backup.py:230:            shutil.rmtree(os.path.join(self.nodes[2].datadir, self.chain, 'chainstate'))
test/functional/wallet_backup.py:251:            os.path.join(self.nodes[0].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename),
test/functional/wallet_backup.py:252:            os.path.join(self.nodes[0].datadir, self.chain, '.', 'wallets', self.default_wallet_name, self.wallet_data_filename),
test/functional/wallet_backup.py:253:            os.path.join(self.nodes[0].datadir, self.chain, 'wallets', self.default_wallet_name),
test/functional/wallet_backup.py:254:            os.path.join(self.nodes[0].datadir, self.chain, 'wallets')]
test/functional/wallet_descriptor.py:237:        wallet_db = os.path.join(self.nodes[0].datadir, self.chain, "wallets", "crashme", self.wallet_data_filename)
test/functional/wallet_hd.py:90:        shutil.rmtree(os.path.join(self.nodes[1].datadir, self.chain, "blocks"))
test/functional/wallet_hd.py:91:        shutil.rmtree(os.path.join(self.nodes[1].datadir, self.chain, "chainstate"))
test/functional/wallet_hd.py:94:            os.path.join(self.nodes[1].datadir, self.chain, 'wallets', self.default_wallet_name, self.wallet_data_filename),
test/functional/wallet_hd.py:118:        shutil.rmtree(os.path.join(self.nodes[1].datadir, self.chain, "blocks"))
test/functional/wallet_hd.py:119:        shutil.rmtree(os.path.join(self.nodes[1].datadir, self.chain, "chainstate"))
test/functional/wallet_hd.py:122:            os.path.join(self.nodes[1].datadir, self.chain, "wallets", self.default_wallet_name, self.wallet_data_filename),
test/functional/wallet_keypool_topup.py:36:        wallet_path = os.path.join(self.nodes[1].datadir, self.chain, "wallets", self.default_wallet_name, self.wallet_data_filename)
test/functional/wallet_listtransactions.py:237:        wallet0 = os.path.join(self.nodes[0].datadir, self.chain, self.default_wallet_name, "wallet.dat")
test/functional/wallet_listtransactions.py:238:        wallet2 = os.path.join(self.nodes[2].datadir, self.chain, self.default_wallet_name, "wallet.dat")
test/functional/wallet_multiwallet.py:65:        data_dir = lambda *p: os.path.join(node.datadir, self.chain, *p)
test/functional/wallet_pruning.py:109:        return os.path.isfile(os.path.join(self.nodes[1].datadir, self.chain, "blocks", f"blk{block_index:05}.dat"))
test/functional/wallet_reorgsrestore.py:92:        shutil.copyfile(os.path.join(self.nodes[0].datadir, 'wallet.bak'), os.path.join(self.nodes[1].datadir, self.chain, self.default_wallet_name, self.wallet_data_filename))
```

...although, changing some of these occurences does require additional changes for example when watching logs.

Are we looking to change all of these out and switch to the os.Path operator syntax throughout?

Commit:
test: Allow pathlib.Path as RPC argument via authproxy

Also, add datadir_path property to TestNode

willcl-ark commented:
Light crACK fa7f1251f8ce4e8cd130194f6d1c6e3ba6aad346

LGTM. Do we want
Looks good to me. There are still some instances outside of _test_framework/_ where we are using `os.path.join` on `self.datadir`. These could also be addressed here or in a followup?

This scripted diff _almost_ works in addressing a few more (but fails inside `with` context manager), but my `sed`-fu is not strong enough to overcome

```sh
sed -i 's/os\.path\.join(self\.nodes\[0\]\.datadir, \(.*\))/self.nodes[0].datadir_path \/ \1/g' test/functional/*.py
```

willcl-ark commented:
Light crACK fa7f1251f8

Looks good to me. 

There are still instances in tests (outside of _test_framework/_) where we are using `os.path.join` on `self.datadir` which we could also switch to using `self.datadir_path` and the `/` operator. Is that something that we are looking to do here or in the future?

theStack reviewed (commented):
LGTM, found only one more instance where `chain_path` could be used:

```diff
diff --git a/test/functional/feature_anchors.py b/test/functional/feature_anchors.py
index 468ad1eafa..0961f21a40 100755
--- a/test/functional/feature_anchors.py
+++ b/test/functional/feature_anchors.py
@@ -20,9 +20,7 @@ class AnchorsTest(BitcoinTestFramework):
         self.disable_autoconnect = False

     def run_test(self):
-        node_anchors_path = os.path.join(
-            self.nodes[0].datadir, "regtest", "anchors.dat"
-        )
+        node_anchors_path = self.nodes[0].chain_path / "anchors.dat"

         self.log.info("When node starts, check if anchors.dat doesn't exist")
         assert not os.path.exists(node_anchors_path)

```

As a follow-up, a new `TestNode` property `wallets_path` might also make sense, to deduplicate even more?

willcl-ark commented:
Same, looks good here. Will ack if you want to make the anchors change.

Question: Now that we use more `Path`s, would there be appetite to replace many of the `os.path.join()` calls with the `/` operator? It looks like ~100 lines or thereabouts, and not any performance wins, nor perfectly scripted-diff-able; so quite a lot of code churn (and might be considered "many grains of sand", and not a nice fat "rock"). But would make the tests more readable IMO?

Commit:
scripted-diff: Use wallets_path and chain_path where possible

Instead of passing the datadir and chain name to os.path.join, just use
the existing properties, which are the same.

-BEGIN VERIFY SCRIPT-
 sed -i --regexp-extended 's|\.datadir, self\.chain, .wallets.|.wallets_path|g' $(git grep -l '\.datadir, self\.chain,')
 sed -i --regexp-extended 's|\.datadir, self\.chain,|.chain_path,|g'            $(git grep -l '\.datadir, self\.chain,')
-END VERIFY SCRIPT-

theStack commented on test/functional/test_framework/authproxy.py:
```diff
@@ -62,6 +63,8 @@ def __init__(self, rpc_error, http_status=None):
 def EncodeDecimal(o):
     if isinstance(o, decimal.Decimal):
         return str(o)
+    if isinstance(o, pathlib.Path):
+        return str(o)
```
nit (feel free to ignore for now, probably follow-up material): Now that this function is not handling decimals only anymore, I think it should be renamed from `EncodeDecimal` to something generic?

fanquake replied:
See #28011.

theStack reviewed (approved):
Code-review ACK aaaa3aefbdfca1c9243057eeefdc19940e60bf18 🌊 

Re-reviewed the changes since my previous review via `$ git range-diff faf830e3...aaaa3aef`

willcl-ark commented:
re-ACK aaaa3aefbdfca1c9243057eeefdc19940e60bf18

LGTM with the wallets_path added.
